### ISSUE TITLE
Add multiple parents when using nested admins

### ISSUE THREAD

**Issue by webdevilopers** [sonata-project/SonataAdminBundle#2402]
So far there can only be one parent set in the `$parentAssociationMapping` attribute of an admin class.

@RobMasters has posted a workaround for this here:
http://stackoverflow.com/questions/10605548/sonata-admin-bundle-possible-to-add-a-child-admin-object-that-can-have-differen

Maybe there is a way to accomplish this similar to the `addChild` method inside the admin class  services.

Any thoughts?

BTW: is this the correct bundle to post the feature request or should I use the https://github.com/sonata-project/SonataDoctrineORMAdminBundle ?

**Comment by webdevilopers** [comment]
Works fine @huitiemesens .

**Comment by webdevilopers** [comment]
When using autocompleters or plural names the hack still needs more workaround for instance:
```php
    public function getParentAssociationMapping()
    {
        // Prevent calling `getObject` on non existing parent when used for autocompletion
        if (!$this->getParent()) {
            return;
        }

        $em = $this->modelManager->getEntityManager('Plusquam\Bundle\ContractBundle\Entity\TimekeepingEntry');
        $className = $em->getClassMetadata(get_class($this->getParent()->getObject($this->getParent()->getRequest()->get('id'))))->getTableName();

        switch ($className)
        {
            case 'users':
                return 'user';
                break;

            case 'contracts':
                return 'contract';
                break;

            default:
                return strtolower( $className );
                break;
        }
        return strtolower( $className );
    }
```

**Comment by lushc** [comment]
I couldn't rely on table names as per @huitiemesens & @webdevilopers approach, so I slightly modified it to use class metadata instead:

```php
public function getParentAssociationMapping()
{
  // use the mapping that has already been set
  if (!is_null($this->parentAssociationMapping)) {
    return $this->parentAssociationMapping;
  }

  // prevent calling getObject on a non-existent parent
  if (!($parent = $this->getParent())) {
    return;
  }

  // get class metadata for the current admin's subject
  $metadata = $this->getModelManager()->getMetadata(get_class($this->getNewInstance()));

  // check if this admin's subject has a field association for the given parent's subject
  if (!$metadata || !($associations = $metadata->getAssociationsByTargetClass(get_class($parent->getNewInstance())))) {
    return;
  }

  // use the first association as the field to access the parent object
  return $this->parentAssociationMapping = strtolower(key($associations));
}
```

**Comment by gremo** [comment]
What about this approach?

```php
/**
 * {@inheritdoc}
 */
public function getParentAssociationMapping()
{
    $mm = $this->getModelManager();
    if ($mm instanceof ModelManager) {
        $associations = $mm->getMetadata($this->getClass())
            ->getAssociationsByTargetClass($this->getParent()->getClass());
        foreach ($associations as $association) {
            switch ($association['type']) {
                case ClassMetadataInfo::MANY_TO_ONE:
                case ClassMetadataInfo::ONE_TO_ONE:
                    return $association['fieldName'];
                default:
                    continue;
            }
        }
    }

    return null;
}
```

**Comment by jordisala1991** [comment]
This is almost finished here: https://github.com/sonata-project/SonataAdminBundle/pull/4954

If someone wants to carry this one, it should be fairly simple.

**Comment by jordisala1991** [comment]
Closed via #5058
